docs/spec_draft.md

```diff
@@ -1798,6 +1799,92 @@ tensor and produces a `result` tensor.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.custom_call
+
+### Semantics
+
+Invokes exteral code with `inputs` as arguments and produces `results`. The
+exact mechanism is backend-specific. The inputs are described as follows:
+  * `inputs` are passed as agruments to the external API.
+  * `call_target_name` may be used as an identifier for the external code.
+  * `has_side_effect` is true if the custom call has side-effects.
+  * `backend_config` may be used to encode additional backend-specific
+    information.
+  * `api_version` specifies the version of the custom call API.
+  * `called_computations` used to apply functions within the scope of the parent
+    module.
+  * `operand_layouts` specifies the layout of the operand.
+  * `result_layouts` specifies the layout of the result.
+  * `output_operand_aliases` captures the alias relationship of `results` to one of the operands.
+    * `operand_index` denotes the index of the operand `inputs` that relates to `results`.
+    * `operand_tuple_indices` are used to index into the operand at `operand_index`, if it is a tuple/nested tuple.
+    * `output_tuple_indices` are used to index into `results`, if it is a tuple/nested tuple.
```

Comment:
Proposed reformulation:

```
Encapsulates an implementation-defined operation `call_target_name` that takes
`inputs` as well as potentially `called_computations` and produces `results`.

`has_side_effect`, `backend_config` and `api_version` may be used to provide 
additional implementation-defined metadata.
```

Feedback:
  1. Invoking external code is only one possible use of CustomCallOp. More generally, it is used for represent implementation-defined operations, which may not even end up referring to any external code, e.g. because they are decomposed by the compiler.
  2. The bullet points in the Semantics section introduce many concepts, e.g. side effects, versions of APIs, layouts, alias relationships, but don't explain them. We should aim to have much lower level of detail in the spec. I have reformulated most of this to say "may be used to provide additional implementation-defined metadata" and proposed to defer layouts and output_operand_aliasing to #629 and #740.
  3. In the future, I'd like to unify all implementation-defined metadata into a single input (see #741 for details). This is future work - there are no action items within this bullet specifically for this PR.